### PR TITLE
"Manual" Payment Method

### PR DESCRIPTION
**Pull request by Kukks**
This PR introduces various small refactorings around payment methods along with introducing a new payment type called "manual". This payment type can be enabled and configured from the store settings and should enable a number of new use-cases.

Some of these use-cases:

- Accepting cash, cheques, gift vouchers in a physical store
- Accepting wire transfer, local banking quick transfer apps, cash, cheques, gift vouchers sent by mail in an online store
- Accepting a split payment of cash and bitcoin 

Configuration of this new payment type is flexible yet simple enough:

image

Checkout UI: 
image

image

When the setting "Mark payment as confirmed instantly (instead of Paid)" is disabled, the invoice is marked as paid when a manual payment is added, until a user with access to the store confirms it. 
image

Additional screenshots:
image

### CODE REVIEW

**Comment by NicolasDorier** [comment]
Need test

File: BTCPayServer/Services/Invoices/InvoiceRepository.cs [hunk 1]
```diff
@@ -174,11 +174,15 @@ public async Task<InvoiceEntity> CreateInvoiceAsync(string storeId, InvoiceEntit
 
                 foreach (var paymentMethod in invoice.GetPaymentMethods())
                 {
-                    if (paymentMethod.Network == null)
+                    if (paymentMethod.Network == null )
```
Comment (NicolasDorier) [inline BTCPayServer/Services/Invoices/InvoiceRepository.cs]: why a space here

File: BTCPayServer/Services/Rates/CurrencyNameTable.cs [hunk 2]
```diff
@@ -142,7 +142,7 @@ public string DisplayFormatCurrency(decimal value, string currency, bool threeLe
             else
                 return value.ToString("C", provider) + $" ({currency})";
         }
-
+
```
Comment (NicolasDorier) [inline BTCPayServer/Services/Rates/CurrencyNameTable.cs]: why

**Comment by NicolasDorier** [comment]
mmmh this is super error prone to have an "empty value" cryptocode. I am wondering if the best is not to move `CryptoCode` out of BTCPayNetworkBase and into BTCPayNetwork. This would make more sense.

**Comment by NicolasDorier** [comment]
Ready for another round of review?

**Comment by Kukks** [comment]
I would say so @NicolasDorier  :)

**Comment by Kukks** [comment]
Good to note that the empty cryptoCode refactoring should be done in a separate PR( and not in #912 but after as it still does not fit with this specific refactoring)

**Comment by NicolasDorier** [comment]
So replace  `string.Empty` with a constant `const string ManualCryptoCode = String.Empty.`, so we can easily cleanup by removing the constant later.

File: BTCPayServer/Payments/PaymentMethodId.cs [hunk 3]
```diff
@@ -76,15 +76,22 @@ public string ToPrettyString()
         public static bool TryParse(string str, out PaymentMethodId paymentMethodId)
         {
             paymentMethodId = null;
-            var parts = str.Split('_', StringSplitOptions.RemoveEmptyEntries);
```
Comment (NicolasDorier) [inline BTCPayServer/Payments/PaymentMethodId.cs]: why do you modify this code?

File: BTCPayServer/Views/Invoice/InvoicePaymentsPartial.cshtml [hunk 4]
```diff
@@ -1,101 +1,49 @@
...
-        </div>
-    </div>
+@{
+    var grouped = Model.Payments.GroupBy(payment => payment.GetPaymentMethodId().PaymentType);
+    
 }
-@if (Model.OffChainPayments.Count > 0)
+@foreach (var paymentGroup in grouped)
 {
-    <div class="row">
```
Comment (NicolasDorier) [inline BTCPayServer/Views/Invoice/InvoicePaymentsPartial.cshtml]: I think this complicate things for nothing. The partial is not really needed as there is no reuse.

File: BTCPayServer/Controllers/InvoiceController.cs [hunk 5]
```diff
@@ -247,11 +248,23 @@ private async Task<PaymentMethod> CreatePaymentMethodAsync(Dictionary<CurrencyPa
                 var logPrefix = $"{supportedPaymentMethod.PaymentId.ToPrettyString()}:";
                 var storeBlob = store.GetStoreBlob();
                 var preparePayment = handler.PreparePayment(supportedPaymentMethod, store, network);
-                var rate = await fetchingByCurrencyPair[new CurrencyPair(network.CryptoCode, entity.ProductInformation.Currency)];
-                if (rate.BidAsk == null)
+                RateResult rate;
+                if (network.CryptoCode == BTCPayNetworkProvider.ManualCryptoCode)
```
Comment (NicolasDorier) [inline BTCPayServer/Controllers/InvoiceController.cs]: Should be unecessary, `network,CryptoCode == entity.ProductInfo.Currency` I should return 1 already

Comment (Kukks) [reply BTCPayServer/Controllers/InvoiceController.cs]: Necessary  as `network.CryptoCode` = "" so `network,CryptoCode != entity.ProductInfo.Currency`

File: BTCPayServer/Views/Invoice/InvoicePaymentsPartial.cshtml (hunk 4 above)
Comment (Kukks) [reply BTCPayServer/Views/Invoice/InvoicePaymentsPartial.cshtml]: I honestly think  it cleans up the code clutter in this section by offloading them to individual sections but I can move them

Comment (NicolasDorier) [reply BTCPayServer/Views/Invoice/InvoicePaymentsPartial.cshtml]: Maybe we can do later, but this PR is already huge and super painful to review. Make it as small as possible so we can move on.

**Comment by NicolasDorier** [comment]
1. I think that the manual payment should NOT have its own "network". This is not a network, this is a payment type.
2. The CryptoCode to empty is super smelly, nor should it be empty. a `PaymentID` of `USD_MANUAL` or `BTC_MANUAL` must be possible. I prefer that we fix the assumption in places where such cryptoCode can always get a BTCPayNetworkProvider rather than hack by saying "empty crypto code", which would completely fuck up the export for example. (The Paid column of the export currency should be what is in the "CryptoCode" column in all case)

**Comment by woutersamaey** [comment]
Can’t we call this “offline” payment or “other” instead of “manual” with a bag of coins icon? Looks confusing on the checkout page. I’d drop out of that checkout.

**Comment by Kukks** [comment]
Superceded by #971

**Pull request closed without merging**